### PR TITLE
BLD: require `pythran>=0.14.0`

### LINKED ISSUE #19733: BUG: `pythran` min version not enforced
### Describe your issue.

Despite `pythran>=0.14.0` in `pyproject.toml`, `python dev.py build` lets me build with `pythran=0.13.1`. This is not true for `Cython`, where the `>=3.0.4` bound is enforced.

### Reproducing Code Example

```python
python dev.py build # with pythran=0.13.1
```

### Error message

```shell
# a wall of warnings which were fixed, hence we bumped the min version
```

### SciPy/NumPy/Python version and system information

<details>

```shell
1.13.0.dev0+1030.3b33204 1.25.2 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /Users/lucascolley/mambaforge/envs/scipy-dev/include
    lib directory: /Users/lucascolley/mambaforge/envs/scipy-dev/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=0 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK=0 NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=1 VORTEX MAX_THREADS=128
    pc file directory: /Users/lucascolley/mambaforge/envs/scipy-dev/lib/pkgconfig
    version: 0.3.23
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /Users/lucascolley/mambaforge/envs/scipy-dev/include
    lib directory: /Users/lucascolley/mambaforge/envs/scipy-dev/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=0 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK=0 NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=1 VORTEX MAX_THREADS=128
    pc file directory: /Users/lucascolley/mambaforge/envs/scipy-dev/lib/pkgconfig
    version: 0.3.23
  pybind11:
    detection method: pkgconfig
    include directory: /Users/lucascolley/mambaforge/envs/scipy-dev/include
    name: pybind11
    version: 2.11.1
Compilers:
  c:
    args: -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong, -O2, -pipe, -isystem,
      /Users/lucascolley/mambaforge/envs/scipy-dev/include, -D_FORTIFY_SOURCE=2, -isystem,
      /Users/lucascolley/mambaforge/envs/scipy-dev/include
    commands: arm64-apple-darwin20.0.0-clang
    linker: ld64
    linker args: -Wl,-pie, -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs,
      -Wl,-rpath,/Users/lucascolley/mambaforge/envs/scipy-dev/lib, -L/Users/lucascolley/mambaforge/envs/scipy-dev/lib,
      -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong, -O2, -pipe, -isystem,
      /Users/lucascolley/mambaforge/envs/scipy-dev/include, -D_FORTIFY_SOURCE=2, -isystem,
      /Users/lucascolley/mambaforge/envs/scipy-dev/include
    name: clang
    version: 15.0.7
  c++:
    args: -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong, -O2, -pipe, -stdlib=libc++,
      -fvisibility-inlines-hidden, -fmessage-length=0, -isystem, /Users/lucascolley/mambaforge/envs/scipy-dev/include,
      -D_FORTIFY_SOURCE=2, -isystem, /Users/lucascolley/mambaforge/envs/scipy-dev/include
    commands: arm64-apple-darwin20.0.0-clang++
    linker: ld64
    linker args: -Wl,-pie, -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs,
      -Wl,-rpath,/Users/lucascolley/mambaforge/envs/scipy-dev/lib, -L/Users/lucascolley/mambaforge/envs/scipy-dev/lib,
      -ftree-vectorize, -fPIC, -fPIE, -fstack-protector-strong, -O2, -pipe, -stdlib=libc++,
      -fvisibility-inlines-hidden, -fmessage-length=0, -isystem, /Users/lucascolley/mambaforge/envs/scipy-dev/include,
      -D_FORTIFY_SOURCE=2, -isystem, /Users/lucascolley/mambaforge/envs/scipy-dev/include
    name: clang
    version: 15.0.7
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.4
  fortran:
    args: -march=armv8.3-a, -ftree-vectorize, -fPIC, -fno-stack-protector, -O2, -pipe,
      -isystem, /Users/lucascolley/mambaforge/envs/scipy-dev/include
    commands: /Users/lucascolley/mambaforge/envs/scipy-dev/bin/arm64-apple-darwin20.0.0-gfortran
    linker: ld64
    linker args: -Wl,-pie, -Wl,-headerpad_max_install_names, -Wl,-dead_strip_dylibs,
      -Wl,-rpath,/Users/lucascolley/mambaforge/envs/scipy-dev/lib, -L/Users/lucascolley/mambaforge/envs/scipy-dev/lib,
      -march=armv8.3-a, -ftree-vectorize, -fPIC, -fno-stack-protector, -O2, -pipe,
      -isystem, /Users/lucascolley/mambaforge/envs/scipy-dev/include
    name: gcc
    version: 12.3.0
  pythran:
    include directory: ../../../mambaforge/envs/scipy-dev/lib/python3.10/site-packages/pythran
    version: 0.13.1
Machine Information:
  build:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
  cross-compiled: false
  host:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
Python Information:
  path: /Users/lucascolley/mambaforge/envs/scipy-dev/bin/python3.10
  version: '3.10'
```

</details>

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
Closes gh-19733

#### What does this implement/fix?
The minimum version for `pythran` is set as `0.14.0` in `pyproject.toml`, but nothing stops you from building with a lesser version, whereas e.g. the minimum bound for `Cython` is enforced.

#### Additional information
I've checked locally that the error is triggered when trying to build with `pythran=0.13.1`, then goes away after `mamba install pythran=0.14.0`.

cc @rgommers

### CODE REVIEW

**Comment by lucascolley** [comment]
ah, looks like we need to upgrade the versions that some of the CI workflows are using

**Comment by lucascolley** [comment]
CI error is gh-19737

**Review by tylerjereddy — APPROVED** [review]
LGTM

**Comment by rgommers** [comment]
Thanks Lucas.& Tyler. IIRC I originally left this out for a reason (maybe an irregular Pythran version string?), but probably that is resolved now. If not, we'll find out soon enough:)

**Comment by eli-schwartz** [comment]
See #19856, the pythran version check can be slightly simpler.

File: .github/workflows/linux_meson.yml [hunk 1]
```diff
@@ -71,7 +71,7 @@ jobs:
       run: |
         python -m pip install git+https://github.com/numpy/numpy.git
         python -m pip install ninja cython pytest pybind11 pytest-xdist pytest-timeout click rich-click doit pydevtool pooch hypothesis "setuptools<67.3"
-        python -m pip install git+https://github.com/serge-sans-paille/pythran.git@e3babfd
+        python -m pip install git+https://github.com/serge-sans-paille/pythran.git
```
Comment (h-vetinari) [inline .github/workflows/linux_meson.yml:74]: Why don't we switch back to released versions here, now that 0.15 is out?

Comment (lucascolley) [reply .github/workflows/linux_meson.yml:74]: Do we need to bump `pythran` to `>=0.15` in `pyproject.toml` too?

Comment (rgommers) [reply .github/workflows/linux_meson.yml:74]: Not just yet, it's only been out for a few days and there are CI failures involving Pythran right now.

Comment (h-vetinari) [reply .github/workflows/linux_meson.yml:74]: Also, aside from some setuptools interaction that I don't know in detail, pythran 0.14 is actually plenty for building scipy.

Comment (lucascolley) [reply .github/workflows/linux_meson.yml:74]: @h-vetinari don't we want to keep this as is in this 'install from repositories' job? `pythran` has been done like this for at least 2 years, see https://github.com/scipy/scipy/blame/8d846f19ac89a2940faa6a356044dc12acc89ac7/.github/workflows/linux_meson.yml.

Comment (h-vetinari) [reply .github/workflows/linux_meson.yml:74]: Fair point. Out of the two occurrences of `.../pythran.git` in https://github.com/scipy/scipy/commit/d1da380d8e357df1d5f8e24558f334c5593eca92, one got already removed, and the other one was pre-existing. As long as we removed the work-arounds introduced between pythran 0.14 & 0.15, I'm fine with leaving things as-is.

**Pull request merged**